### PR TITLE
Document another limitation of driver-only hashes

### LINKED ISSUE #6497: X.509 with PSS fails with accelerated hash
If you have a PSA hash accelerator, you disable the software implementation, you enable `MBEDTLS_USE_PSA_CRYPTO` and don't disable `MBEDTLS_MD_C`, then X.509 certificates and CRL using RSA-PSS cannot be verified. The problem is a lack of consistency in determining which hash algorithms are supported. It ultimately rests on e.g. `MBEDTLS_SHA256_C` in code that dispatches through legacy APIs and on `PSA_WANT_ALG_SHA_256` in code that dispatches through PSA, and we don't consistently choose one or the other.

`legacy_or_psa.h` as of today (383b0bbea08aefdc81833a11a5425fe4e2eba130, which is what we're planning to release in 3.3 as far as this feature is concerned) does not resolve this problem. In X.509, availability of the hash is determined by `MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA_BASED_ON_USE_PSA`. But when it gets down to the RSA code, it instead uses the presence of `MBEDTLS_MD_C` to determine whether to call PSA or legacy.

How to reproduce: I wanted to modify an `all.sh` component (when looking for redundancies) to have one that tests a build where some but not all hashes are accelerated. It's a variant of `component_test_psa_crypto_config_accel_hash_use_psa`.
```
component_test_psa_crypto_config_accel_hash_except_sha256 () {
    msg "test: MBEDTLS_PSA_CRYPTO_CONFIG with accelerated hash except SHA256"

    # Disable ALG_STREAM_CIPHER and ALG_ECB_NO_PADDING to avoid having
    # partial support for cipher operations in the driver test library.
    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_STREAM_CIPHER
    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_ECB_NO_PADDING

    loc_accel_list="ALG_MD5 ALG_RIPEMD160 ALG_SHA_1 ALG_SHA_224 ALG_SHA_384 ALG_SHA_512"
    loc_accel_flags=$( echo "$loc_accel_list" | sed 's/[^ ]* */-DLIBTESTDRIVER1_MBEDTLS_PSA_ACCEL_&/g' )
    make -C tests libtestdriver1.a CFLAGS="$ASAN_CFLAGS $loc_accel_flags" LDFLAGS="$ASAN_CFLAGS"

    # start with config full for maximum coverage (also enables USE_PSA)
    scripts/config.py full
    scripts/config.py set MBEDTLS_PSA_CRYPTO_DRIVERS
    scripts/config.py set MBEDTLS_PSA_CRYPTO_CONFIG
    scripts/config.py unset MBEDTLS_MD5_C
    scripts/config.py unset MBEDTLS_RIPEMD160_C
    scripts/config.py unset MBEDTLS_SHA1_C
    # Don't unset MBEDTLS_SHA256_C. It will be used for entropy.
    scripts/config.py unset MBEDTLS_SHA384_C
    scripts/config.py unset MBEDTLS_SHA512_C
    scripts/config.py unset MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT
    loc_accel_flags="$loc_accel_flags $( echo "$loc_accel_list" | sed 's/[^ ]* */-DMBEDTLS_PSA_ACCEL_&/g' )"
    make CFLAGS="$ASAN_CFLAGS -Werror -I../tests/include -I../tests -I../../tests -DPSA_CRYPTO_DRIVER_TEST -DMBEDTLS_TEST_LIBTESTDRIVER1 $loc_accel_flags" LDFLAGS="-ltestdriver1 $ASAN_CFLAGS"

    not grep mbedtls_sha512_init library/sha512.o
    not grep mbedtls_sha1_init library/sha1.o

    msg "test: MBEDTLS_PSA_CRYPTO_CONFIG with accelerated hash"
    make test
}
```

In 383b0bbea08aefdc81833a11a5425fe4e2eba130 this fails in `test_suite_x509parse`:
```
…
X509 CRT information RSA-PSS, SHA1 Digest ......................... FAILED
  strcmp( buf, result_str ) == 0
  at line 478, suites/test_suite_x509parse.function
X509 CRT information RSA-PSS, SHA224 Digest ....................... PASS
X509 CRT information RSA-PSS, SHA256 Digest ....................... PASS
X509 CRT information RSA-PSS, SHA384 Digest ....................... FAILED
  strcmp( buf, result_str ) == 0
  at line 478, suites/test_suite_x509parse.function
X509 CRT information RSA-PSS, SHA512 Digest ....................... FAILED
  strcmp( buf, result_str ) == 0
  at line 478, suites/test_suite_x509parse.function
…
X509 CRL information RSA-PSS, SHA1 Digest ......................... FAILED
  strcmp( buf, result_str ) == 0
  at line 501, suites/test_suite_x509parse.function
X509 CRL information RSA-PSS, SHA224 Digest ....................... PASS
X509 CRL information RSA-PSS, SHA256 Digest ....................... PASS
X509 CRL information RSA-PSS, SHA384 Digest ....................... FAILED
  strcmp( buf, result_str ) == 0
  at line 501, suites/test_suite_x509parse.function
X509 CRL information RSA-PSS, SHA512 Digest ....................... FAILED
  strcmp( buf, result_str ) == 0
  at line 501, suites/test_suite_x509parse.function
…
X509 CSR Information RSA-PSS with SHA1 ............................ FAILED
  strcmp( buf, result_str ) == 0
  at line 540, suites/test_suite_x509parse.function
X509 CSR Information RSA-PSS with SHA224 .......................... PASS
X509 CSR Information RSA-PSS with SHA256 .......................... PASS
X509 CSR Information RSA-PSS with SHA384 .......................... FAILED
  strcmp( buf, result_str ) == 0
  at line 540, suites/test_suite_x509parse.function
X509 CSR Information RSA-PSS with SHA512 .......................... FAILED
  strcmp( buf, result_str ) == 0
  at line 540, suites/test_suite_x509parse.function
…
X509 CRT verification #57 (Valid, RSASSA-PSS, SHA-1) .............. FAILED
  res == ( result )
  at line 671, suites/test_suite_x509parse.function
X509 CRT verification #58 (Valid, RSASSA-PSS, SHA-224) ............ PASS
X509 CRT verification #59 (Valid, RSASSA-PSS, SHA-256) ............ PASS
X509 CRT verification #60 (Valid, RSASSA-PSS, SHA-384) ............ FAILED
  res == ( result )
  at line 671, suites/test_suite_x509parse.function
X509 CRT verification #61 (Valid, RSASSA-PSS, SHA-512) ............ FAILED
  res == ( result )
  at line 671, suites/test_suite_x509parse.function
X509 CRT verification #62 (Revoked, RSASSA-PSS, SHA-1) ............ FAILED
  flags == (uint32_t)( flags_result )
  at line 672, suites/test_suite_x509parse.function
X509 CRT verification #63 (Revoked, RSASSA-PSS, SHA-1, CRL badsign  FAILED
  flags == (uint32_t)( flags_result )
  at line 672, suites/test_suite_x509parse.function
X509 CRT verification #64 (Valid, RSASSA-PSS, SHA-1, not top) ..... FAILED
  res == ( result )
  at line 671, suites/test_suite_x509parse.function
X509 CRT verification #65 (RSASSA-PSS, SHA1, bad cert signature) .. PASS
X509 CRT verification #66 (RSASSA-PSS, SHA1, no RSA CA) ........... PASS
X509 CRT verification #67 (Valid, RSASSA-PSS, all defaults) ....... FAILED
  res == ( result )
  at line 671, suites/test_suite_x509parse.function
…
```
The test dependencies uses the macros for hash support in X.509, but they need to also require hash support in PSS. In `mbedtls-3.2.1`, this passes because the test cases depend on `MBEDTLS_SHAx_C`, which is overly restrictive.

It's an easy fix for the dependencies of these test cases, but it highlights a design problem, so this is more justification for https://github.com/Mbed-TLS/mbedtls/issues/6471.

I have not looked for other similar mismatches that might not be tested.

### PR DESCRIPTION
**Pull request by mpg**
## Description

Fixes #6497 by documenting the limitation.

## Gatekeeper checklist

- [ ] **changelog** not required - edited an existing entry
- [ ] **backport** not required - new feature
- [ ] **tests** not required - documentation only

### CODE REVIEW

File: ChangeLog.d/driver-only-hashes.txt [hunk 1]
```diff
@@ -8,12 +8,16 @@ Features
...
-     features still depend on the built-in implementation of hashes:
-     MBEDTLS_HKDF_C (but the PSA HKDF function do not depend on it),
-     MBEDTLS_ENTROPY_C, MBEDTLS_HMAC_DRBG_C and MBEDTLS_ECDSA_DETERMINISTIC.
-     In particular, for now, compiling without built-in hashes requires use
-     of MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG.
+     implementation is still available. Note that for modules that use MD
+     (RSA, PKCS5, PKCS12, EC J-PAKE) in builds that have MBEDTLS_MD_C enabled,
+     all hashes used with those modules need to be built-in, as drivers are only
+     used when MBEDTLS_MD_C is disabled; configurations where some hashes are
+     available as built-ins, and some only from drivers, are currently not
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/driver-only-hashes.txt:15]: This seems overly restrictive. What we don't support is a mix of hashes that are only available via the legacy interface but not PSA, and others that are only available as PSA drivers. If all hashes are available through the PSA interface, it's fine, regardless of whether they're provided via drivers or not.

File: ChangeLog.d/driver-only-hashes.txt [hunk 2]
```diff
@@ -8,12 +8,16 @@ Features
...
      modules; this is not required in configurations where the built-in
-     implementation is still available. Note that some crypto modules and
-     features still depend on the built-in implementation of hashes:
-     MBEDTLS_HKDF_C (but the PSA HKDF function do not depend on it),
-     MBEDTLS_ENTROPY_C, MBEDTLS_HMAC_DRBG_C and MBEDTLS_ECDSA_DETERMINISTIC.
-     In particular, for now, compiling without built-in hashes requires use
-     of MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG.
+     implementation is still available. Note that for modules that use MD
+     (RSA, PKCS5, PKCS12, EC J-PAKE) in builds that have MBEDTLS_MD_C enabled,
+     all hashes used with those modules need to be built-in, as drivers are only
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/driver-only-hashes.txt:13]: Another example of a long sentence that's hard to follow: “modules that use MD (…) in builds that have MBEDTLS_MD_C enabled” — the context and the beginning makes the reader think this is about the case with MBEDTLS_MD_C disabled, but it's about the case where it is enabled. It would be easier to follow if the sentence started with “When MBEDTLS_MD_C is enabled”.

File: ChangeLog.d/driver-only-hashes.txt [hunk 3]
```diff
@@ -8,12 +8,16 @@ Features
...
+     used when MBEDTLS_MD_C is disabled; configurations where some hashes are
+     available as built-ins, and some only from drivers, are currently not
+     supported. Also note that some crypto modules and features still depend on
+     the built-in implementation of hashes: MBEDTLS_HKDF_C (but the PSA HKDF
+     functions do not depend on it), MBEDTLS_ENTROPY_C, MBEDTLS_HMAC_DRBG_C and
+     MBEDTLS_ECDSA_DETERMINISTIC. In particular, for now, compiling without
+     built-in hashes requires use of MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/driver-only-hashes.txt:20]: Except for builds without entropy (e.g. only doing hashes, MAc, symmetric decryption, and signature verification).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The changelog entry needs work

File: ChangeLog.d/driver-only-hashes.txt (hunk 2 above)
Comment (mpg) [reply ChangeLog.d/driver-only-hashes.txt:13]: In line with what I said above, a lot of sentences would automatically get a lot simpler if we moved the info to `mbedtls_config.h` and described only one module at a time.

File: ChangeLog.d/driver-only-hashes.txt (hunk 3 above)
Comment (gilles-peskine-arm) [reply ChangeLog.d/driver-only-hashes.txt:20]: Ah, good point, you can't yet have PSA hashes without having a RNG.

File: ChangeLog.d/driver-only-hashes.txt [hunk 4]
```diff
@@ -8,12 +8,16 @@ Features
      are only provided by PSA drivers. In these configurations, you need to
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/driver-only-hashes.txt:8]: We can put line break wherever we want. Changelog entries aren't rewrapped.

I think the deeper problem here is that the feature we're delivering doesn't really make sense and should not have been a deliverable in this state. But it's too late for that, so we have to document what we have. Or at least, we have to document a subset of what we have — we can be incomplete if it's just too hard to explain. We expect the next release to have much simpler, broader rules.

File: ChangeLog.d/driver-only-hashes.txt [hunk 5]
```diff
@@ -1,20 +1,19 @@
...
-     MBEDTLS_ENTROPY_C, MBEDTLS_HMAC_DRBG_C and MBEDTLS_ECDSA_DETERMINISTIC.
-     In particular, for now, compiling without built-in hashes requires use
-     of MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG.
-   * When MBEDTLS_USE_PSA_CRYPTO is enabled, X.509, TLS 1.2 and TLS 1.3 no
-     longer depend on MD. This means it is now possible to use them in
-     configurations where the built-in implementations of hashes are excluded
-     and the hashes are only provided by PSA drivers.
+   * Some modules can now use PSA drivers for hashes, including with no
+     built-in implementation present, but only in some configurations.
+     - RSA PKCS#1 v2.1, PKCS5, PKCS12 and EC J-PAKE now use hashes from PSA
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/driver-only-hashes.txt:4]: Not everyone can be expected to know that “PKCS#1 v2.1” means OAEP and PSS. (And strictly speaking, it doesn't: the v1.5 mechanisms are still included in the v2.1 document.)
```suggestion
     - RSA (PSS and OAEP), PKCS5, PKCS12 and EC J-PAKE now use hashes from PSA
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Thanks for the thorough rewrite! Looks good to me apart from one minor quibble.

File: include/mbedtls/mbedtls_config.h [hunk 6]
```diff
@@ -1146,6 +1146,11 @@
  * \warning If building without MBEDTLS_MD_C, you must call psa_crypto_init()
  * before doing any PKCS#1 v2.1 operation.
  *
+ * \warning When building with MBEDTLS_MD_C, all hashes used with this
+ * need to be available a built-ins (that is, for SHA-256, MBEDTLS_SHA256_C,
```
Comment (aditya-deshpande-arm) [inline include/mbedtls/mbedtls_config.h:1150]: [_Typo_] Did you mean "available **as** built-ins"? Same goes for paragraphs below.

**Review by aditya-deshpande-arm — COMMENTED** [review]
LGTM

File: ChangeLog.d/driver-only-hashes.txt (hunk 5 above)
Comment (mpg) [reply ChangeLog.d/driver-only-hashes.txt:4]: Good point. Though PKCS#1 v2.1 is the name of the compile-time option, so perhaps it's worth keeping in addition to PSS and OAEP?

(Also, strictly speaking, we could say we're referring to the version that introduced those schemes, but even that wouldn't be correct: OAEP was introduced in 2.0 and PSS in 2.1...)

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

File: ChangeLog.d/driver-only-hashes.txt (hunk 5 above)
Comment (gilles-peskine-arm) [reply ChangeLog.d/driver-only-hashes.txt]: The name of the compile-time option is `MBEDTLS_PKCS1_V21`, not “PKCS#1 v2.1”. But as long as PSS and OAEP are listed, that's fine by me.

**Review by aditya-deshpande-arm — APPROVED** [review]
